# CICERO 0.3.0

<div align="center">

  [![Status](https://img.shields.io/badge/status-active-success.svg)](https://github.com/stjude/CICERO)
  [![GitHub issues](https://img.shields.io/github/issues/stjude/CICERO)](https://github.com/stjude/CICERO/issues)
  [![GitHub Pull Requests](https://img.shields.io/github/issues-pr/stjude/CICERO)](https://github.com/stjude/CICERO/pulls)

</div>

CICERO (Clipped-reads Extended for RNA Optimization) is an assembly-based algorithm to detect diverse classes
of driver gene fusions from RNA-seq.

<br />
<p align="center">
  <img alt="Overview of CICERO algorithm which consists of fusion detection through analysis of candidate SV breakpoints and splice junction, fusion annotation, and ranking." src="docs/CICERO.png"/>
</p>
<br />

## üìù Table of Contents

- [Running CICERO](#running)
- [Dependencies](#dependencies)
- [Running with Docker](#docker)
- [Running with St. Jude Cloud](#cloud)
- [Generate Junctions](#junctions)
- [Reference Files](#reference)
- [Supported Genomes](#genomes)
- [Demo](#demo)
- [Citation](#citation)
- [License](#license)

## Running CICERO <a name="running"></a>

Add the `src/scripts` directory to your system `PATH` variable. Add the `src/perllib` and `dependencies/lib/perl` directories to your system `PERL5LIB` variable.

Then invoke the CICERO wrapper as

```bash
Cicero.sh [-n ncores] -b bamfile -g genome -r refdir -o outdir [-j junctions] [-t threshold] [-s sc_cutoff] [-c sc_shift] [-p]
```

- `ncores` is the number of cores to be run on (with [GNU parallel](https://www.gnu.org/software/parallel/)).
- `bamfile` is the input bamfile mapped to human genome builds GRCh37-lite or GRCh38_no_alt. [Contact us](mailto:liqing.tian@stjude.org) if your bam is based on other reference version.
- `genome` is either `GRCh37-lite` or `GRCh38_no_alt`. CICERO only support the two human reference genome versions.
- `refdir` is the reference file directory specific to CICERO. Download [Reference Files](#reference) below. e.g. "-r /home/users/softwares/CICERO/reference_hg38/" or "-r /home/users/softwares/CICERO/reference_hg19/"
- `outdir` is user defined output file folder.
- `junctions` is the junctions file output from RNApeg. See [Generate Junctions](#junctions) below. CICERO can detect fusion by analysis of splice junction reads. If this option is omitted, fusions generated by small deletions may be missed as these events may lack the soft-clipped reads.
- `threshold` CICERO first detects all soft-clipped positions supported by >=2 reads from bam file. For sample with <=threshold (default 200,000) soft-clipped positions, CICERO will detect fusions based on these soft-clipped positons; otherwise, to speed-up CICERO running, CICERO will detect fusions based on soft-clipped positions supported by >=sc_cutoff (3, default for optimize mode, see blew) reads. For sample with lots of soft-clipped positions, a smaller threshold will speed-up CICERO running, however, some fusion events (i.e. only supported by 2 reads) may be missed.
- `sc_cutoff` controls the number of soft clip reads required to support a putative site. The default is 2, but for samples with large numbers of soft clip reads, it may be desirable to require additional support to reduce the computational time required.
- `sc_shift` sets the threshold for considering events to be the same site.
- `optimize` defaults to ON. This sets `sc_cutoff` to 3 for samples where the number of soft clip sites exceeds 200,000. It also sets `sc_shift` to 10 which sets the distance to consider events the same.
- `-no-optimize` turns optimizations off. This can increase sensitivity, but increases the computational requirements.

The final CICERO fusion result file will be located at `<outdir>/CICERO_DATADIR/<sample name>/final_fusions.txt`. Use the following [guide](https://www.stjude.cloud/docs/guides/genomics-platform/analyzing-data/rapid-rnaseq/) to interpret the results.

To visualize CICERO fusion output you can load the final fusion output file at https://proteinpaint.stjude.org/FusionEditor/.

## Dependencies <a name="dependencies"></a>

* [GNU parallel](https://www.gnu.org/software/parallel/)
* [Samtools 1.3.1](http://www.htslib.org/doc/samtools-1.3.1.html)
* [Cap3](https://www.ncbi.nlm.nih.gov/pubmed/10508846)
* [Blat](https://genome.ucsc.edu/goldenpath/help/blatSpec.html)
* Java 1.8.0
* Perl 5.10.1 with libraries:
    - base
    - Bio
    - Carp
    - Compress
    - Cwd
    - Data
    - DBI
    - diagnostics
    - Digest
    - English
    - enum
    - Exporter
    - File
    - FileHandle
    - List
    - POSIX
    - strict
    - Sys
    - Tree
    - warnings

## Running with Docker <a name="docker"></a>

CICERO can be run with Docker. To begin, build the Docker image using the Dockerfile in this repository.

```bash
docker build -t stjude/cicero:0.3.0 .
```

Then invoke the CICERO wrapper using Docker.

```bash
docker run -v <path to reference directory>:/reference stjude/cicero:0.3.0 Cicero.sh [-n cores] -b <bam file path> -g <genome> -r /reference -o <output directory> [-j junctions file] [-p] [-s int] [-t int] [-c int]
```

See [Running CICERO](#running) for details of the parameters.

## Running with St. Jude Cloud <a name="cloud"></a>

CICERO is integrated in the St. Jude Cloud Rapid RNA-Seq workflow. To run CICERO in St. Jude Cloud, access the tool through the [platform page](https://platform.stjude.cloud/workflows/rapid_rna-seq). Documentation
for running and interpreting results is available in the [user guide](https://stjudecloud.github.io/docs/guides/genomics-platform/analyzing-data/rapid-rnaseq/).

## Generate junctions file with RNApeg <a name="junctions"></a>

RNApeg is required to generate a junctions file for use by CICERO. You can get RNApeg from both Docker and Singularity. Once RNApeg is complete, the `*.junctions.tab.shifted.tab` file can be provided to CICERO using the `-j` argument. 

### Running RNApeg via Docker:

```bash
docker run -v <outdir>:/results mnedmonson/public:rnapeg RNApeg.sh -b bamfile -f fasta -r refflat
```

- `fasta` reference genome; i.e. "Homo_sapiens/GRCh38_no_alt/FASTA/GRCh38_no_alt.fa" or "Homo_sapiens/GRCh37-lite/FASTA/GRCh37-lite.fa" from [Reference Files](#reference).
- `refflat` i.e. "Homo_sapiens/GRCh38_no_alt/mRNA/RefSeq/refFlat.txt" or "Homo_sapiens/GRCh37-lite/mRNA/Combined/all_refFlats.txt" from [Reference Files](#reference).

### Running RNApeg via Singularity:

```bash
singularity run --bind <outdir>:/results docker://mnedmonson/public:rnapeg RNApeg.sh -b bamfile -f fasta -r refflat
```

## Downloading reference files <a name="reference"></a>

Reference files are required to run CICERO. They can be found at the following location:

- GRCh37-lite: https://doi.org/10.5281/zenodo.3817656
- GRCh38_no_alt: https://doi.org/10.5281/zenodo.3894739

## Supported Genome Version <a name="genomes"></a>

CICERO currently supports `GRCh37-lite` and `GRCh38_no_alt`. 

## Demo <a name="demo"></a>

A demo of CICERO can be found at the following location:
* https://www.stjuderesearch.org/site/lab/zhang/cicero

## Citation <a name="citation"></a>

Tian, L., Li, Y., Edmonson, M.N. et al. CICERO: a versatile method for detecting complex and diverse driver fusions using cancer RNA sequencing data. Genome Biol 21, 126 (2020). https://doi.org/10.1186/s13059-020-02043-x

## License <a name="license"></a>

Copyright 2019 St. Jude Children's Research Hospital

Licensed under a modified version of the Apache License, Version 2.0
(the "License") for academic research use only; you may not use this
file except in compliance with the License. To inquire about commercial
use, please contact the St. Jude Office of Technology Licensing at
scott.elmer@stjude.org.

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
